packages/backend-function/src/provided_function_factory.ts

```diff
@@ -0,0 +1,143 @@
+  BackendOutputStorageStrategy,
+  ConstructContainerEntryGenerator,
+  ConstructFactory,
+  ConstructFactoryGetInstanceProps,
+  FunctionResources,
+  GenerateContainerEntryProps,
+  ResourceAccessAcceptor,
+} from '@aws-amplify/plugin-types';
+import { Construct } from 'constructs';
+import { CfnFunction, IFunction } from 'aws-cdk-lib/aws-lambda';
+import { FunctionOutput } from '@aws-amplify/backend-output-schemas';
+import { Tags } from 'aws-cdk-lib';
+import { AmplifyUserError, TagName } from '@aws-amplify/platform-core';
+import { AmplifyFunctionBase } from './function_construct_base.js';
+import { FunctionResourceAccessAcceptor } from './resource_access_acceptor.js';
+
+export type ProvidedFunctionProps = {
+  /**
+   * Group the function with existing Amplify resources or separate the function into its own group.
+   * @default 'function' // grouping with other Amplify functions
+   * @example
+   * resourceGroupName: 'auth' // to group an auth trigger with an auth resource
+   */
+  resourceGroupName?: AmplifyResourceGroupName;
+};
+
+/**
+ * Adapts provided CDK function as Amplify function.
+ */
+export class ProvidedFunctionFactory
+  implements ConstructFactory<AmplifyFunction>
+{
+  private generator: ConstructContainerEntryGenerator;
+
+  /**
+   * Creates provided function factory.
+   */
+  constructor(
+    private readonly functionProvider: (scope: Construct) => IFunction,
+    private readonly props?: ProvidedFunctionProps
+  ) {}
+
+  /**
+   * Creates a function instance.
+   */
+  getInstance(props: ConstructFactoryGetInstanceProps): AmplifyFunction {
+    if (!this.generator) {
+      this.generator = new ProvidedFunctionGenerator(
+        this.functionProvider,
+        props.outputStorageStrategy,
+        this.props
+      );
+    }
+    return props.constructContainer.getOrCompute(
+      this.generator
+    ) as AmplifyFunction;
+  }
+}
+
+class ProvidedFunctionGenerator implements ConstructContainerEntryGenerator {
+  readonly resourceGroupName: AmplifyResourceGroupName;
+
+  constructor(
+    private readonly functionProvider: (scope: Construct) => IFunction,
+    private readonly outputStorageStrategy: BackendOutputStorageStrategy<FunctionOutput>,
+    props?: ProvidedFunctionProps
+  ) {
+    this.resourceGroupName = props?.resourceGroupName ?? 'function';
+  }
+
+  generateContainerEntry = ({ scope }: GenerateContainerEntryProps) => {
+    let providedFunction: IFunction;
+    try {
+      providedFunction = this.functionProvider(scope);
+    } catch (e) {
+      if (
+        e instanceof Error &&
+        (e.message.includes('docker exited with status 1') ||
+          e.message.includes('docker ENOENT'))
+      ) {
+        throw new AmplifyUserError(
+          'FunctionBundlingDockerError',
+          {
+            message: e.message,
+            // TODO better resolution
+            resolution:
+              'Ensure that docker is present and works correctly. See https://some.link.about.docker.',
```

Comment:
We can use go functions without and error, an also we can use python functions without docker. I tried and gave example before how we can use python without docker Any change to pass at least for golang?

Reply:
What about we make bundling without docker on Linux by default in python alpha package? On ci/cd it will use Amazon Linux, so we don't need docker in Linux for Python, it can be fast solution

Reply:
> What about we make bundling without docker on Linux by default in python alpha package? On ci/cd it will use Amazon Linux, so we don't need docker in Linux for Python, it can be fast solution

These kind of improvements can be added later outside of this PR. What we need to merge this PR is docs that clearly state the scope of support. Please allow us a bit more time to come up with that.

Reply:
I can also create docs for creating new go and python runtimes with this new support, is this works or should I wait?

Reply:
> I can also create docs for creating new go and python runtimes with this new support, is this works or should I wait?

You could give it a stab. We should aim to have a sub page under https://docs.amplify.aws/react/build-a-backend/functions/ . Our docs are defined in https://github.com/aws-amplify/docs repo.

Couple of things to keep in mind:
1. The page should be called "Custom functions"
2. The page must call out lack of docker support.
3. We can use docker-less python build as an example.
4. I recommend to avoid GO as an example (it doesn't seem to have first class support, see https://aws.amazon.com/blogs/compute/migrating-aws-lambda-functions-from-the-go1-x-runtime-to-the-custom-runtime-on-amazon-linux-2/ )
5. We might need to wait for reviewers quorum until after Jan 6th due to holidays.

Reply:
Okay I'll try it, Golang is not anymore directly runtime but it still can be using directly just by giving runtime as provided.al2023, I personally using this without an error, I would be happy if I can add Golang there with Python.

Reply:
I created PR for documents [here](https://github.com/aws-amplify/docs/pull/8176). Happy holidays.

Reply:
@josefaidt please take a look at docs PR ^ when you're back.